smart_campus/app/views.py

```diff
@@ -1,3 +1,149 @@
 from django.shortcuts import render
+from django.views.decorators.csrf import csrf_exempt
+from django.views.decorators.http import require_POST
+from django.contrib.auth import authenticate, login, logout
+from django.http import HttpResponseRedirect, JsonResponse
+from django.contrib.auth.decorators import login_required
+from django.contrib import messages
 
-# Create your views here.
+from .models import (
+    User, Reward, Permission,
+    Station,
+)
+
+
+@csrf_exempt
+@require_POST
+def signup(request):
+    """Signup API for APP users
+
+    Register new users
+
+    """
+    email = request.POST.get('email')
+    password = request.POST.get('password')
+    nickname = request.POST.get('nickname')
+
+    if User.objects.filter(email=email).exists():
+        return JsonResponse({'status': 'false', 'message': 'User already exists!'})
+
+    try:
+        User.objects.create_user(email, password, nickname)
+    except ValueError as error:
+        return JsonResponse({'status': 'false', 'message': error})
+
+    return JsonResponse({'status': 'true', 'message': 'Success'})
+
+
+@csrf_exempt
+@require_POST
+def signin(request):
+    """Sign in API for APP users
+
+    Handle signin requests from app
+
+    """
+    email = request.POST.get('email')
+    password = request.POST.get('password')
+    user = authenticate(request, username=email, password=password)
+    if user is not None:
+        login(request, user)
+        user_data = {
+            'nickname': user.nickname,
+            'experience_point': user.experience_point,
+            'coins': user.earned_coins,
+            'reward': [reward.id for reward in user.received_rewards.all()],
+            'favorite_stations': [station.id for station in user.favorite_stations.all()],
+        }
+        return JsonResponse({'status': 'true', 'message': 'Success', 'data': user_data})
+
+    return JsonResponse({'status': 'false', 'message': 'Login failed'})
+
+
+@csrf_exempt
+@require_POST
+def signout(request):
+    """Sign out API for APP users
+
+    Handle signout requests from app
+
+    """
+    logout(request)
+    return JsonResponse({'status': 'true', 'message': 'Success'})
+
+
+@csrf_exempt
+def login_page(request):
+    """Login for Management Backend Page
+
+    Render Login page and handle login requests for the Management backend
+
+    """
+    if request.user.is_authenticated():
+        return HttpResponseRedirect('/')
+
+    email = request.POST.get('email')
+    password = request.POST.get('password')
+    user = authenticate(request, username=email, password=password)
+
+    if user is not None and user.can(Permission.EDIT):
+        login(request, user)
+        return HttpResponseRedirect('/')
+
+    if user is None and email is not None:
```

Comment:
If I'm not mistaken, this line is used to check whether the password is correct.
It's confusing.
I spend some time to figure it out.
You can consider refactoring the code or at least write some comment and refactor it once you have time.

Reply:
I suggest that you can remove the `email is no None` check and move the code before the ones that check user permission.

Reply:
Password check is done in 'authenticate', which returns None if the email and pwd is not correct.

Reply:
The reason I'm doing line 93 is just for the message.
When someone tries to login but failed. 
Expression 'user is None and email is not None' will be True

Reply:
Loading the login page will have all 3(email, password, user ) to be None

Reply:
The sequence of the code block is a little bit weird.
What do you think of the following sequence?

```python
    if not user:
        messages.warning(request, 'Login failed!')
        return render(request, 'app/login.html')

    if user.can(Permission.EDIT) or user.can(Permission.ADMIN):
        # Although it can be long if you add future permissions, this can save a SQL query
    	# which might be more efficient.
    	# This check is also reasonable since we're not sure whether the user with future
    	# added permissions can be logged in.
        login(request, user)
        return HttpResponseRedirect('/')
```

Reply:
Oh, because we share the same view with render and handle login request.
So, at the very first time a user opens the page, he'll get the 'login failed' message.

To deal with this,
I've made the form field to be required so that I am able to distinguish whether it's a login request(email and pwd will not be None) or just a first time page load.

Reply:
The sequence is nice, but the 'Login failed' message will show.

Reply:
Agree with the permission checks.

Reply:
Thinking of another way, maybe I could check whether it's a POST or GET request ?!

Reply:
Yep, check the request method is a common way to solve this problem.
Great idea!